File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task170.java
import java.util.Arrays;

public class Task170 {

    /**
     * Given an integer array nums and two integers lower and upper, return the number of range sums
     * that lie in [lower, upper] inclusive.
     *
     * @param nums   The input integer array.
     * @param lower  The lower bound of the range sum.
     * @param upper  The upper bound of the range sum.
     * @return The number of range sums in the inclusive range [lower, upper].
     */
    public int countRangeSum(int[] nums, int lower, int upper) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int n = nums.length;
        // Prefix sums can exceed Integer.MAX_VALUE, so use long.
        long[] prefixSums = new long[n + 1];
        for (int i = 0; i < n; i++) {
            prefixSums[i + 1] = prefixSums[i] + nums[i];
        }
        return mergeSortAndCount(prefixSums, 0, n, lower, upper);
    }

    /**
     * A recursive helper function using a divide and conquer approach (Merge Sort).
     * It counts the valid range sums and sorts the prefix sum subarray.
     *
     * @param sums   The array of prefix sums.
     * @param start  The starting index of the subarray (inclusive).
     * @param end    The ending index of the subarray (inclusive).
     * @param lower  The lower bound of the target range.
     * @param upper  The upper bound of the target range.
     * @return The count of valid range sums within this subarray and across its halves.
     */
    private int mergeSortAndCount(long[] sums, int start, int end, int lower, int upper) {
        if (start >= end) {
            return 0;
        }
        int mid = start + (end - start) / 2;
        int count = mergeSortAndCount(sums, start, mid, lower, upper) +
                    mergeSortAndCount(sums, mid + 1, end, lower, upper);

        // Count range sums S(i, j) where i is in the left half and j in the right half.
        // For each sums[j] in the right half, we need to find how many sums[i] in the left half
        // satisfy: lower <= sums[j] - sums[i] <= upper
        // which is equivalent to: sums[j] - upper <= sums[i] <= sums[j] - lower
        int k = start, l = start;
        for (int j = mid + 1; j <= end; j++) {
            while (k <= mid && sums[k] < sums[j] - upper) {
                k++;
            }
            while (l <= mid && sums[l] <= sums[j] - lower) {
                l++;
            }
            count += l - k;
        }

        // Standard merge step for merge sort
        long[] temp = new long[end - start + 1];
        int i = start, j = mid + 1, t = 0;
        while (i <= mid && j <= end) {
            if (sums[i] <= sums[j]) {
                temp[t++] = sums[i++];
            } else {
                temp[t++] = sums[j++];
            }
        }
        while (i <= mid) {
            temp[t++] = sums[i++];
        }
        while (j <= end) {
            temp[t++] = sums[j++];
        }
        System.arraycopy(temp, 0, sums, start, temp.length);
        
        return count;
    }

    public static void main(String[] args) {
        Task170 solution = new Task170();

        // Test Case 1
        int[] nums1 = {-2, 5, -1};
        int lower1 = -2, upper1 = 2;
        System.out.println("Test Case 1: " + solution.countRangeSum(nums1, lower1, upper1));

        // Test Case 2
        int[] nums2 = {0};
        int lower2 = 0, upper2 = 0;
        System.out.println("Test Case 2: " + solution.countRangeSum(nums2, lower2, upper2));

        // Test Case 3
        int[] nums3 = {1, 1, 1};
        int lower3 = 1, upper3 = 2;
        System.out.println("Test Case 3: " + solution.countRangeSum(nums3, lower3, upper3));

        // Test Case 4
        int[] nums4 = {-1, -1, -1};
        int lower4 = -2, upper4 = -1;
        System.out.println("Test Case 4: " + solution.countRangeSum(nums4, lower4, upper4));

        // Test Case 5
        int[] nums5 = {2147483647, -2147483648, -1, 0};
        int lower5 = -1, upper5 = 0;
        System.out.println("Test Case 5: " + solution.countRangeSum(nums5, lower5, upper5));
    }
}

File: dataset/CoT/GEMINI_gemini-2.5-pro/Task170.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int mergeSortAndCount(long long* sums, int start, int end, int lower, int upper) {
    if (start >= end) {
        return 0;
    }

    int mid = start + (end - start) / 2;
    int count = mergeSortAndCount(sums, start, mid, lower, upper) + 
                mergeSortAndCount(sums, mid + 1, end, lower, upper);

    int l = mid + 1;
    int r = mid + 1;
    for (int i = start; i <= mid; ++i) {
        while (l <= end && sums[l] - sums[i] < lower) {
            l++;
        }
        while (r <= end && sums[r] - sums[i] <= upper) {
            r++;
        }
        count += (r - l);
    }
    
    // Merge step
    int tempSize = end - start + 1;
    long long* temp = (long long*)malloc(tempSize * sizeof(long long));
    if (temp == NULL) {
        exit(1); 
    }
    
    int p1 = start, p2 = mid + 1, p = 0;
    while (p1 <= mid && p2 <= end) {
        if (sums[p1] <= sums[p2]) {
            temp[p++] = sums[p1++];
        } else {
            temp[p++] = sums[p2++];
        }
    }
    while (p1 <= mid) {
        temp[p++] = sums[p1++];
    }
    while (p2 <= end) {
        temp[p++] = sums[p2++];
    }

    for (int i = 0; i < tempSize; ++i) {
        sums[start + i] = temp[i];
    }
    
    free(temp);

    return count;
}

int countRangeSum(int* nums, int numsSize, int lower, int upper) {
    if (nums == NULL || numsSize == 0) {
        return 0;
    }
    
    long long* prefixSums = (long long*)malloc((numsSize + 1) * sizeof(long long));
    if (prefixSums == NULL) {
        exit(1);
    }
    
    prefixSums[0] = 0;
    for (int i = 0; i < numsSize; ++i) {
        prefixSums[i + 1] = prefixSums[i] + nums[i];
    }
    
    int result = mergeSortAndCount(prefixSums, 0, numsSize, lower, upper);
    
    free(prefixSums);
    
    return result;
}

int main() {
    // Test Case 1
    int nums1[] = {-2, 5, -1};
    int lower1 = -2, upper1 = 2;
    printf("Test Case 1: %d\n", countRangeSum(nums1, sizeof(nums1)/sizeof(int), lower1, upper1));

    // Test Case 2
    int nums2[] = {0};
    int lower2 = 0, upper2 = 0;
    printf("Test Case 2: %d\n", countRangeSum(nums2, sizeof(nums2)/sizeof(int), lower2, upper2));

    // Test Case 3
    int nums3[] = {1, -1, 1, -1};
    int lower3 = 0, upper3 = 0;
    printf("Test Case 3: %d\n", countRangeSum(nums3, sizeof(nums3)/sizeof(int), lower3, upper3));

    // Test Case 4
    int nums4[] = {INT_MAX, INT_MIN, 0, -1, 1};
    int lower4 = -1, upper4 = 1;
    printf("Test Case 4: %d\n", countRangeSum(nums4, sizeof(nums4)/sizeof(int), lower4, upper4));
    
    // Test Case 5
    int nums5[] = {0, 0, 0, 0, 0};
    int lower5 = 0, upper5 = 0;
    printf("Test Case 5: %d\n", countRangeSum(nums5, sizeof(nums5)/sizeof(int), lower5, upper5));
    
    return 0;
}